dash-pipeline/utils/dash_pipeline_utils.py

```diff
@@ -0,0 +1,361 @@
+        action = entry.action.action
+
+        param = action.params.add()
+        param.param_id = 1
+        if neighbor_mac:
+            param.value = neighbor_mac
+        else:   # default value
+            param.value = b'\x00\x00\x00\x00\x00\x00'
+
+        param = action.params.add()
+        param.param_id = 2
+        if mac:
+            param.value = mac
+        else:   # default value
+            param.value = b'\x00\x00\x00\x00\x00\x00'
+
+        param = action.params.add()
+        param.param_id = 3
+        if cpu_mac:
+            param.value = cpu_mac
+        else:   # default value
+            param.value = b'\x00\x00\x00\x00\x00\x00'
+
+        param = action.params.add()
+        param.param_id = 4
+        if flow_enabled:
+            param.value = flow_enabled
+        else:   # default value
+            param.value = b'\x00'
+
+        self.write(entry, p4runtime_pb2.Update.INSERT)
+
+
+class P4FlowTable(P4Table):
+    def __init__(self, target=None):
+        super(P4FlowTable, self).__init__(target)
+        self.p4info_table_flow = self.p4info.get_table("dash_ingress.conntrack_lookup_stage.flow_entry")
+
+    def print_flow_table(self):
+        for entry in self.read(self.p4info_table_flow.preamble.id):
+            print(entry)
+
+    def get_flow_entry(self, eni_mac, vnet_id,
+                       src_ip, dst_ip,
+                       src_port, dst_port, ip_proto):
+        match_list = []
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 1
+        match.exact.value = mac_in_bytes(eni_mac)
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 2
+        match.exact.value = vnet_id.to_bytes(2, byteorder='big')
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 3
+        match.exact.value = ip_address(src_ip).packed
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 4
+        match.exact.value = ip_address(dst_ip).packed
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 5
+        match.exact.value = src_port.to_bytes(2, byteorder='big')
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 6
+        match.exact.value = dst_port.to_bytes(2, byteorder='big')
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 7
+        match.exact.value = ip_proto.to_bytes(1, byteorder='big')
+        match_list.append(match)
+
+        match = p4runtime_pb2.FieldMatch()
+        match.field_id = 8
+        match.exact.value = b'\x00' if ip_address(src_ip).version == 4 else b'\x01'
+        match_list.append(match)
+
+        for entry in self.read(self.p4info_table_flow.preamble.id, match_list):
+            return entry
+
+        return None
+
+
+def use_flow(cls):
```

Comment:
maybe we should move these 2 functions to a dedicated file, since these 2 functions will not work for pytest.

Reply:
Looks pytest supports unittest-based test. Will check if it may support pytest

Reply:
Still keep use_flow in p4_dash_utils.py